In [1]:
import requests
from bs4 import BeautifulSoup

ModuleNotFoundError: No module named 'bs4'

In [ ]:
import pandas as pd
import json
import numpy as np

In [ ]:
gameName='Hitman 3'
gameNameAlt='Hitman 3'

---- allkeyshop search----- 

In [ ]:
url_endpoint = "https://www.allkeyshop.com/blog/wp-admin/admin-ajax.php"
reqParams = {
    'action': 'catalogue_query',
    'query[search]': gameName,
    'query[productsPerPage]': 16,
    'query[currency]': 'eur',
    'query[category]': 'pc-games-all'
}

In [ ]:
resp = requests.get(url_endpoint, params=reqParams).json()['results']
resp

'    <li class="search-results-row">\n    <a href="https://www.allkeyshop.com/blog/buy-hitman-3-cd-key-compare-prices/" class="search-results-row-link">\n        <div class="search-results-row-image">\n            <div class="aks-lazy search-results-row-image-ratio" style="background-image: url(https://www.allkeyshop.com/img/buy-hitman-3-cd-key-pc-download-catalog.jpg)">\n            </div>\n        </div>\n        <div class="search-results-row-game">\n            <h2 class="search-results-row-game-title">Hitman 3</h2>\n            <div class="search-results-row-game-infos">2021 - Action</div>\n        </div>\n        <div class="metacritic d-none d-xl-block">\n<div class="metacritic-button metacritic-button-basic metacritic-button-green" data-metacritic data-product-id="38888">\n    8.0    <div class="metacritic-spinner-wrapper">\n        <div class="spinner spinner-20"></div>\n    </div>\n</div>\n</div>\n        <div class="search-results-row-price">\n                            15.

In [ ]:
soup = BeautifulSoup(resp, 'html.parser')
productId = soup.find(class_="metacritic-button")['data-product-id']
productId

'38888'

In [ ]:
reqParams = {
    'action': 'get_offers',
    'product': productId,
    'currency': 'eur',
    'use_beta_offers_display': 1,
    'region':2
}

payload = {}

response = requests.request(
    "GET", url_endpoint, params=reqParams, data=payload)
response.json()

{'success': True,
 'offers': [{'id': 134513555,
   'affiliateUrl': 'https://www.eneba.com/steam-hitman-3-pc-steam-key-global?af_id=allkeyshop',
   'isActive': True,
   'merchant': '272',
   'price': {'eur': {'bestCoupon': {'code': '8%: AKS8',
      'discountValue': '8.00',
      'discountStrategy': '%',
      'isCashback': False},
     'currency': 'eur',
     'price': 18.38,
     'priceWithoutCoupon': 19.98}},
   'edition': '1',
   'region': '2',
   'stock': 'InStock',
   'platform': 'steam'},
  {'id': 134524098,
   'affiliateUrl': 'https://www.kinguin.net/category/115956/hitman-3-steam-cd-key?r=3445&nosalesbooster=1&currency=EUR&discount=OCTOBER14',
   'isActive': True,
   'merchant': '47',
   'price': {'eur': {'bestCoupon': {'code': '14%: OCTOBER14',
      'discountValue': '14.00',
      'discountStrategy': '%',
      'isCashback': False},
     'currency': 'eur',
     'price': 23.83,
     'priceWithoutCoupon': 27.71}},
   'edition': '1',
   'region': '2',
   'stock': 'InStock',
   'p

In [ ]:
#f=open('test.json','r')
#file=json.loads(f.read())
file=response.json()

In [ ]:
file.keys()

dict_keys(['success', 'offers', 'merchants', 'editions', 'regions'])

In [ ]:
file['offers'][0]['price']['eur'].keys()

dict_keys(['bestCoupon', 'currency', 'price', 'priceWithoutCoupon'])

In [ ]:
file['offers'][0]['price']['eur']['priceWithoutCoupon']

19.98

In [ ]:
data=pd.json_normalize(file,record_path='offers')
data

id                                       affiliateUrl  isActive  \
0   134513555  https://www.eneba.com/steam-hitman-3-pc-steam-...      True   
1   134524098  https://www.kinguin.net/category/115956/hitman...      True   
2   133071217  https://www.gamivo.com/product/hitman-3?glv=ki...      True   
3   134652039  https://www.instant-gaming.com/en/12619-/?igr=...      True   
4   134569366  https://www.cdkeys.com/hitman-3-pc-steam?mw_ar...      True   
5   134524473  https://www.hrkgame.com/en/games/product/hitma...      True   
6   134533054  https://www.cjs-cdkeys.com/products/HITMAN-3-S...      True   
7   134597584  https://k4g.com/product/hitman-3-steam-global-...      True   
8   134518514  https://www.g2a.com/hitman-3-pc-steam-key-glob...      True   
9   134451227  https://royalcdkeys.com/products/hitman-3-stea...      True   
10  134518513  https://www.g2a.com/hitman-3-pc-steam-key-glob...      True   
11  134652550  https://www.play-asia.com/hitman-iii/13/70fltx...      True   
12  134689466  https://store.steampowered.com/app/1659040/HIT...      True   
13  134576917  http://www.gamingdragons.com/en/game/buy-hitma...      True   
14  134556470  https://www.kinguin.net/category/118902/hitman...      True   
15  134545660  https://www.gamivo.com/product/hitman-3-deluxe...      True   
16  134638685  https://www.hrkgame.com/en/games/product/hitma...      True   
17  134668487  https://k4g.com/product/hitman-3-steam-global-...      True   
18  134514503  https://www.eneba.com/steam-hitman-3-deluxe-ed...      True   
19  134523817  https://www.g2a.com/hitman-3-deluxe-edition-pc...      True   
20  134523816  https://www.g2a.com/hitman-3-deluxe-edition-pc...      True   
21  134658897  https://www.voidu.com/en/product/hitman-3-stan...      True   
22  134658895  https://www.voidu.com/en/product/hitman-3-delu...      True   
23  134567518  https://www.gamers-outlet.net/buy-hitman-3-cd-...      True   
24  134634311  https://www.cdkeys.com/hitman-3-deluxe-edition...      True   

   merchant edition region       stock platform price.eur.bestCoupon.code  \
0       272       1      2     InStock    steam                  8%: AKS8   
1        47       1      2     InStock    steam            14%: OCTOBER14   
2       218       1      2     InStock    steam          8%: AKSpooktober   
3        13       1      2     InStock    steam                       NaN   
4         9       1      2     InStock    steam                       NaN   
5        94       1      2     InStock    steam        -10%: ALLKEYSHOP10   
6        67       1      2     InStock    steam                       NaN   
7       324       1      2     InStock    steam         14%: AKS14PUMPKIN   
8     61616       1      2     InStock    steam         1.99€ Monthly Sub   
9       321       1      2     InStock    steam                       NaN   
10       61       1      2     InStock    steam                       NaN   
11       49       1      2     InStock    steam                       NaN   
12        1       1      2     InStock    steam                       NaN   
13       73       1      2     InStock    steam                       NaN   
14       47       7      2     InStock    steam            14%: OCTOBER14   
15      218       7      2     InStock    steam          8%: AKSpooktober   
16       94       7      2     InStock    steam        -10%: ALLKEYSHOP10   
17      324       7      2     InStock    steam         14%: AKS14PUMPKIN   
18      272       7      2     InStock    steam                  8%: AKS8   
19    61616       7      2     InStock    steam         1.99€ Monthly Sub   
20       61       7      2     InStock    steam                       NaN   
21      222       1      2     InStock    steam                       NaN   
22      222       7      2     InStock    steam                       NaN   
23       77       1      2  OutOfStock    steam                       NaN   
24        9       7      2  OutOfStock    steam          

mapp=data.drop(['price.eur.bestCoupon.discountValue','price.eur.bestCoupon.discountStrategy','price.eur.bestCoupon.isCashback','price.eur.bestCoupon.code', 'price.eur.bestCoupon'], axis=1,inplace=True)

In [ ]:
data

id                                       affiliateUrl  isActive  \
0   134513555  https://www.eneba.com/steam-hitman-3-pc-steam-...      True   
1   134524098  https://www.kinguin.net/category/115956/hitman...      True   
2   133071217  https://www.gamivo.com/product/hitman-3?glv=ki...      True   
3   134652039  https://www.instant-gaming.com/en/12619-/?igr=...      True   
4   134569366  https://www.cdkeys.com/hitman-3-pc-steam?mw_ar...      True   
5   134524473  https://www.hrkgame.com/en/games/product/hitma...      True   
6   134533054  https://www.cjs-cdkeys.com/products/HITMAN-3-S...      True   
7   134597584  https://k4g.com/product/hitman-3-steam-global-...      True   
8   134518514  https://www.g2a.com/hitman-3-pc-steam-key-glob...      True   
9   134451227  https://royalcdkeys.com/products/hitman-3-stea...      True   
10  134518513  https://www.g2a.com/hitman-3-pc-steam-key-glob...      True   
11  134652550  https://www.play-asia.com/hitman-iii/13/70fltx...      True   
12  134689466  https://store.steampowered.com/app/1659040/HIT...      True   
13  134576917  http://www.gamingdragons.com/en/game/buy-hitma...      True   
14  134556470  https://www.kinguin.net/category/118902/hitman...      True   
15  134545660  https://www.gamivo.com/product/hitman-3-deluxe...      True   
16  134638685  https://www.hrkgame.com/en/games/product/hitma...      True   
17  134668487  https://k4g.com/product/hitman-3-steam-global-...      True   
18  134514503  https://www.eneba.com/steam-hitman-3-deluxe-ed...      True   
19  134523817  https://www.g2a.com/hitman-3-deluxe-edition-pc...      True   
20  134523816  https://www.g2a.com/hitman-3-deluxe-edition-pc...      True   
21  134658897  https://www.voidu.com/en/product/hitman-3-stan...      True   
22  134658895  https://www.voidu.com/en/product/hitman-3-delu...      True   
23  134567518  https://www.gamers-outlet.net/buy-hitman-3-cd-...      True   
24  134634311  https://www.cdkeys.com/hitman-3-deluxe-edition...      True   

   merchant edition region       stock platform price.eur.bestCoupon.code  \
0       272       1      2     InStock    steam                  8%: AKS8   
1        47       1      2     InStock    steam            14%: OCTOBER14   
2       218       1      2     InStock    steam          8%: AKSpooktober   
3        13       1      2     InStock    steam                       NaN   
4         9       1      2     InStock    steam                       NaN   
5        94       1      2     InStock    steam        -10%: ALLKEYSHOP10   
6        67       1      2     InStock    steam                       NaN   
7       324       1      2     InStock    steam         14%: AKS14PUMPKIN   
8     61616       1      2     InStock    steam         1.99€ Monthly Sub   
9       321       1      2     InStock    steam                       NaN   
10       61       1      2     InStock    steam                       NaN   
11       49       1      2     InStock    steam                       NaN   
12        1       1      2     InStock    steam                       NaN   
13       73       1      2     InStock    steam                       NaN   
14       47       7      2     InStock    steam            14%: OCTOBER14   
15      218       7      2     InStock    steam          8%: AKSpooktober   
16       94       7      2     InStock    steam        -10%: ALLKEYSHOP10   
17      324       7      2     InStock    steam         14%: AKS14PUMPKIN   
18      272       7      2     InStock    steam                  8%: AKS8   
19    61616       7      2     InStock    steam         1.99€ Monthly Sub   
20       61       7      2     InStock    steam                       NaN   
21      222       1      2     InStock    steam                       NaN   
22      222       7      2     InStock    steam                       NaN   
23       77       1      2  OutOfStock    steam                       NaN   
24        9       7      2  OutOfStock    steam          

In [ ]:
data.rename(columns = {'price.eur.currency':'currency', 'price.eur.price':'price','price.eur.priceWithoutCoupon':'priceWithoutCoupon'}, inplace = True)

In [ ]:
data[data.priceWithoutCoupon==data.priceWithoutCoupon.min()]

id                                       affiliateUrl  isActive  \
0  134513555  https://www.eneba.com/steam-hitman-3-pc-steam-...      True   

  merchant edition region    stock platform price.eur.bestCoupon.code  \
0      272       1      2  InStock    steam                  8%: AKS8   

  price.eur.bestCoupon.discountValue price.eur.bestCoupon.discountStrategy  \
0                               8.00                                     %   

  price.eur.bestCoupon.isCashback currency  price  priceWithoutCoupon  \
0                           False      eur  18.38               19.98   

   price.eur.bestCoupon  
0                   NaN

---------- dataframe for merchants ------------

In [ ]:
merchants_list=pd.DataFrame(data=file['merchants']).T
merchants_list

id            name                 aggregateRating           types  \
272      272           Eneba   {'value': 3.62, 'count': 313}    CD Key Store   
47        47         Kinguin   {'value': 4.17, 'count': 586}    CD Key Store   
218      218          GAMIVO  {'value': 4.18, 'count': 1247}    CD Key Store   
13        13  Instant Gaming   {'value': 4.25, 'count': 307}    CD Key Store   
9          9      CDKeys.com   {'value': 4.17, 'count': 724}    CD Key Store   
94        94             HRK   {'value': 2.71, 'count': 321}    CD Key Store   
67        67      CJS-CDKeys    {'value': 3.8, 'count': 812}    CD Key Store   
324      324             K4G   {'value': 4.84, 'count': 673}    CD Key Store   
61616  61616        G2A Plus   {'value': 4.29, 'count': 292}    CD Key Store   
321      321   Royal CD Keys    {'value': 4.4, 'count': 289}    CD Key Store   
61        61             G2A   {'value': 4.29, 'count': 292}    CD Key Store   
49        49       Play-Asia    {'value': 4.55, 'count': 10}    CD Key Store   
1          1           Steam     {'value': 4.8, 'count': 46}  Official Store   
73        73   Gamingdragons    {'value': 3.04, 'count': 69}    CD Key Store   
222      222        Voidu EU   {'value': 3.89, 'count': 100}  Official Store   
77        77   Gamers Outlet   {'value': 4.09, 'count': 151}    CD Key Store   

                                         paymentMethods        logoSlug  \
272              {'paymentMethode': ['card', 'paypal']}           eneba   
47               {'paymentMethode': ['card', 'paypal']}         kinguin   
218    {'paymentMethode': ['card', 'paypal', 'crypto']}          gamivo   
13               {'paymentMethode': ['paypal', 'card']}  instant-gaming   
9                {'paymentMethode': ['card', 'paypal']}      cdkeys-com   
94     {'paymentMethode': ['card', 'paypal', 'crypto']}             hrk   
67               {'paymentMethode': ['card', 'paypal']}      cjs-cdkeys   
324              {'paymentMethode': ['card', 'paypal']}             k4g   
61616            {'paymentMethode': ['card', 'paypal']}             g2a   
321                        {'paymentMethode': ['card']}   royal-cd-keys   
61               {'paymentMethode': ['card', 'paypal']}             g2a   
49               {'paymentMethode': ['card', 'paypal']}       play-asia   
1                {'paymentMethode': ['card', 'paypal']}           steam   
73               {'paymentMethode': ['card', 'paypal']}   gamingdragons   
222              {'paymentMethode': ['card', 'paypal']}           voidu   
77               {'paymentMethode': ['card', 'paypal']}   gamers-outlet   

                                               reviewUrl  
272        https://www.allkeyshop.com/blog/review/eneba/  
47       https://www.allkeyshop.com/blog/review/kinguin/  
218       https://www.allkeyshop.com/blog/review/gamivo/  
13     https://www.allkeyshop.com/blog/review/instant...  
9      https://www.allkeyshop.com/blog/review/cdkeys_...  
94       https://www.allkeyshop.com/blog/review/hrkgame/  
67     https://www.allkeyshop.com/blog/review/cjs-cdk...  
324          https://www.allkeyshop.com/blog/review/k4g/  
61616        https://www.allkeyshop.com/blog/review/g2a/  
321    https://www.allkeyshop.com/blog/review/royalcd...  
61           https://www.allkeyshop.com/blog/review/g2a/  
49     https://www.allkeyshop.com/blog/review/play-asia/  
1          https://www.allkeyshop.com/blog/review/steam/  
73     https://www.allkeyshop.com/blog/review/gamingd...  
222        https://www.allkeyshop.com/blog/review/voidu/  
77     https://www.allkeyshop.com/blog/review/gamers-...

---- merge merchants with games hits to make better names -----

In [ ]:
merged_list = pd.merge(data,merchants_list[['id','name']],left_on='merchant',right_on='id',how='left')
merged_list.drop(['id_y','price'], axis=1,inplace=True)
merged_list.rename(columns = {'priceWithoutCoupon':'price'}, inplace = True)
#merged_list.set_index('id_x',inplace=True)

In [ ]:
merged_list.head(-12)

id_x                                       affiliateUrl  isActive  \
0   134513555  https://www.eneba.com/steam-hitman-3-pc-steam-...      True   
1   134524098  https://www.kinguin.net/category/115956/hitman...      True   
2   133071217  https://www.gamivo.com/product/hitman-3?glv=ki...      True   
3   134652039  https://www.instant-gaming.com/en/12619-/?igr=...      True   
4   134569366  https://www.cdkeys.com/hitman-3-pc-steam?mw_ar...      True   
5   134524473  https://www.hrkgame.com/en/games/product/hitma...      True   
6   134533054  https://www.cjs-cdkeys.com/products/HITMAN-3-S...      True   
7   134597584  https://k4g.com/product/hitman-3-steam-global-...      True   
8   134518514  https://www.g2a.com/hitman-3-pc-steam-key-glob...      True   
9   134451227  https://royalcdkeys.com/products/hitman-3-stea...      True   
10  134518513  https://www.g2a.com/hitman-3-pc-steam-key-glob...      True   
11  134652550  https://www.play-asia.com/hitman-iii/13/70fltx...      True   
12  134689466  https://store.steampowered.com/app/1659040/HIT...      True   

   merchant edition region    stock platform price.eur.bestCoupon.code  \
0       272       1      2  InStock    steam                  8%: AKS8   
1        47       1      2  InStock    steam            14%: OCTOBER14   
2       218       1      2  InStock    steam          8%: AKSpooktober   
3        13       1      2  InStock    steam                       NaN   
4         9       1      2  InStock    steam                       NaN   
5        94       1      2  InStock    steam        -10%: ALLKEYSHOP10   
6        67       1      2  InStock    steam                       NaN   
7       324       1      2  InStock    steam         14%: AKS14PUMPKIN   
8     61616       1      2  InStock    steam         1.99€ Monthly Sub   
9       321       1      2  InStock    steam                       NaN   
10       61       1      2  InStock    steam                       NaN   
11       49       1      2  InStock    steam                       NaN   
12        1       1      2  InStock    steam                       NaN   

   price.eur.bestCoupon.discountValue price.eur.bestCoupon.discountStrategy  \
0                                8.00                                     %   
1                               14.00                                     %   
2                                8.00                                     %   
3                                 NaN                                   NaN   
4                                 NaN                                   NaN   
5                               10.00                                     %   
6                                 NaN                                   NaN   
7                               14.00                                     %   
8                                0.00                                     %   
9                                 NaN                                   NaN   
10                                NaN                                   NaN   
11                                NaN                                   NaN   
12                                NaN                                   NaN   

   price.eur.bestCoupon.isCashback currency  price  price.eur.bestCoupon  \
0                            False      eur  19.98                   NaN   
1                            False      eur  27.71                   NaN   
2                            False      eur  27.88                   NaN   
3                              NaN      eur  25.75                   NaN   
4                              NaN      eur  25.79                   NaN   
5                            False      eur  29.97                   NaN   
6                              NaN      eur  27.05                   NaN   
7                            False      eur  32.08                   NaN   
8                            False      eur  28.20                   NaN   
9           

In [ ]:
instock_list=merged_list.loc[(merged_list['platform'] == 'steam') & (merged_list['stock'] == 'InStock')]

In [ ]:
instock_list['price'].describe()

count    23.000000
mean     41.044783
std      16.402772
min      19.980000
25%      28.040000
50%      33.810000
75%      53.005000
max      89.990000
Name: price, dtype: float64

In [ ]:
table=instock_list[["name", "price"]].groupby("name").min()
table.sort_values('price', ascending=True,inplace=True)

In [ ]:
table.head(5)

price
name                 
Eneba           19.98
Instant Gaming  25.75
CDKeys.com      25.79
CJS-CDKeys      27.05
Kinguin         27.71

In [ ]:
instock_list['price'].min()

19.98

In [ ]:
instock_list[instock_list.price==instock_list.price.min()]

id_x                                       affiliateUrl  isActive  \
0  134513555  https://www.eneba.com/steam-hitman-3-pc-steam-...      True   

  merchant edition region    stock platform price.eur.bestCoupon.code  \
0      272       1      2  InStock    steam                  8%: AKS8   

  price.eur.bestCoupon.discountValue price.eur.bestCoupon.discountStrategy  \
0                               8.00                                     %   

  price.eur.bestCoupon.isCashback currency  price  price.eur.bestCoupon   name  
0                           False      eur  19.98                   NaN  Eneba

In [ ]:
instock_list[instock_list.price==instock_list.price.min()]

id_x                                       affiliateUrl  isActive  \
0  134513555  https://www.eneba.com/steam-hitman-3-pc-steam-...      True   

  merchant edition region    stock platform price.eur.bestCoupon.code  \
0      272       1      2  InStock    steam                  8%: AKS8   

  price.eur.bestCoupon.discountValue price.eur.bestCoupon.discountStrategy  \
0                               8.00                                     %   

  price.eur.bestCoupon.isCashback currency  price  price.eur.bestCoupon   name  
0                           False      eur  19.98                   NaN  Eneba

--- here starts currency collection from ecb-------

In [ ]:
# Building blocks for the URL
entrypoint = 'https://sdw-wsrest.ecb.europa.eu/service/' # Using protocol 'https'
resource = 'data'           # The resource for data queries is always'data'
flowRef ='EXR'              # Dataflow describing the data that needs to be returned, exchange rates in this case
key = 'D.SEK.EUR.SP00.A'    # Defining the dimension values, explained below

# Define the parameters
parameters = {
    'startPeriod': '2000-01-01',  # Start date of the time series
    'endPeriod': '2022-10-10'     # End of the time series
}

In [ ]:
# Construct the URL: https://sdw-wsrest.ecb.europa.eu/service/data/EXR/D.CHF.EUR.SP00.A
request_url = entrypoint + resource + '/'+ flowRef + '/' + key

# Make the HTTP request again, now requesting for CSV format
response = requests.get(request_url, params=parameters, headers={'Accept': 'text/csv'})
# Response succesful? (Response code 200)
print(response)

<Response [200]>


In [ ]:
open('temp.csv', 'wb').write(response.content)
currency_df = pd.read_csv('temp.csv')

In [ ]:
currency_df.keys()

Index(['KEY', 'FREQ', 'CURRENCY', 'CURRENCY_DENOM', 'EXR_TYPE', 'EXR_SUFFIX',
       'TIME_PERIOD', 'OBS_VALUE', 'OBS_STATUS', 'OBS_CONF', 'OBS_PRE_BREAK',
       'OBS_COM', 'TIME_FORMAT', 'BREAKS', 'COLLECTION', 'COMPILING_ORG',
       'DISS_ORG', 'DOM_SER_IDS', 'PUBL_ECB', 'PUBL_MU', 'PUBL_PUBLIC',
       'UNIT_INDEX_BASE', 'COMPILATION', 'COVERAGE', 'DECIMALS', 'NAT_TITLE',
       'SOURCE_AGENCY', 'SOURCE_PUB', 'TITLE', 'TITLE_COMPL', 'UNIT',
       'UNIT_MULT'],
      dtype='object')

In [ ]:
currency_df.tail(1)['OBS_VALUE']

5890    10.9502
Name: OBS_VALUE, dtype: float64

In [ ]:
10.9502*16.69

182.75883800000003

In [ ]:
ts = currency_df.filter(['TIME_PERIOD', 'OBS_VALUE'], axis=1)
# 'TIME_PERIOD' was of type 'object' (as seen in df.info). Convert it to datetime first
ts['TIME_PERIOD'] = pd.to_datetime(ts['TIME_PERIOD'])
# Set 'TIME_PERIOD' to be the index
ts = ts.set_index('TIME_PERIOD')
# Print the last 5 rows to screen
ts.tail(1)

OBS_VALUE
TIME_PERIOD           
2022-10-10     10.9502

In [ ]:
rate_sek = currency_df.tail(1)['OBS_VALUE'].values[0]
rate_sek

10.9502

--- end of currency collection -----

In [ ]:
allkeysSE = instock_list.assign(priceSEK=instock_list["price"]*rate_sek)
#df2.round({'priceSEK': 1})
allkeysSE.priceSEK=allkeysSE.priceSEK.round(2)
allkeysSE.sort_values(by='priceSEK')

id_x                                       affiliateUrl  isActive  \
0   134513555  https://www.eneba.com/steam-hitman-3-pc-steam-...      True   
3   134652039  https://www.instant-gaming.com/en/12619-/?igr=...      True   
4   134569366  https://www.cdkeys.com/hitman-3-pc-steam?mw_ar...      True   
6   134533054  https://www.cjs-cdkeys.com/products/HITMAN-3-S...      True   
1   134524098  https://www.kinguin.net/category/115956/hitman...      True   
2   133071217  https://www.gamivo.com/product/hitman-3?glv=ki...      True   
8   134518514  https://www.g2a.com/hitman-3-pc-steam-key-glob...      True   
5   134524473  https://www.hrkgame.com/en/games/product/hitma...      True   
9   134451227  https://royalcdkeys.com/products/hitman-3-stea...      True   
10  134518513  https://www.g2a.com/hitman-3-pc-steam-key-glob...      True   
7   134597584  https://k4g.com/product/hitman-3-steam-global-...      True   
11  134652550  https://www.play-asia.com/hitman-iii/13/70fltx...      True   
12  134689466  https://store.steampowered.com/app/1659040/HIT...      True   
13  134576917  http://www.gamingdragons.com/en/game/buy-hitma...      True   
15  134545660  https://www.gamivo.com/product/hitman-3-deluxe...      True   
14  134556470  https://www.kinguin.net/category/118902/hitman...      True   
19  134523817  https://www.g2a.com/hitman-3-deluxe-edition-pc...      True   
16  134638685  https://www.hrkgame.com/en/games/product/hitma...      True   
18  134514503  https://www.eneba.com/steam-hitman-3-deluxe-ed...      True   
20  134523816  https://www.g2a.com/hitman-3-deluxe-edition-pc...      True   
17  134668487  https://k4g.com/product/hitman-3-steam-global-...      True   
21  134658897  https://www.voidu.com/en/product/hitman-3-stan...      True   
22  134658895  https://www.voidu.com/en/product/hitman-3-delu...      True   

   merchant edition region    stock platform price.eur.bestCoupon.code  \
0       272       1      2  InStock    steam                  8%: AKS8   
3        13       1      2  InStock    steam                       NaN   
4         9       1      2  InStock    steam                       NaN   
6        67       1      2  InStock    steam                       NaN   
1        47       1      2  InStock    steam            14%: OCTOBER14   
2       218       1      2  InStock    steam          8%: AKSpooktober   
8     61616       1      2  InStock    steam         1.99€ Monthly Sub   
5        94       1      2  InStock    steam        -10%: ALLKEYSHOP10   
9       321       1      2  InStock    steam                       NaN   
10       61       1      2  InStock    steam                       NaN   
7       324       1      2  InStock    steam         14%: AKS14PUMPKIN   
11       49       1      2  InStock    steam                       NaN   
12        1       1      2  InStock    steam                       NaN   
13       73       1      2  InStock    steam                       NaN   
15      218       7      2  InStock    steam          8%: AKSpooktober   
14       47       7      2  InStock    steam            14%: OCTOBER14   
19    61616       7      2  InStock    steam         1.99€ Monthly Sub   
16       94       7      2  InStock    steam        -10%: ALLKEYSHOP10   
18      272       7      2  InStock    steam                  8%: AKS8   
20       61       7      2  InStock    steam                       NaN   
17      324       7      2  InStock    steam         14%: AKS14PUMPKIN   
21      222       1      2  InStock    steam                       NaN   
22      222       7      2  InStock    steam                       NaN   

   price.eur.bestCoupon.discountValue price.eur.bestCoupon.discountStrategy  \
0                                8.00                                     %   
3                                 NaN                                   NaN   
4                                 NaN                                   NaN   
6                                 NaN      

#drop crappy - G2A/plus, Eneba
allkeysSE.drop(allkeysSE[allkeysSE['merchant'] == '61616'].index, inplace = True)
allkeysSE.drop(allkeysSE[allkeysSE['merchant'] == '272'].index, inplace = True)
allkeysSE.drop(allkeysSE[allkeysSE['merchant'] == '61'].index, inplace = True)
allkeysSE.drop(allkeysSE[allkeysSE['merchant'] == '224'].index, inplace = True)
allkeysSE

In [ ]:
allkeysSE.loc[allkeysSE['merchant']=='13',['priceSEK','name']].sort_values(by='priceSEK').head(3)

priceSEK            name
3    281.97  Instant Gaming

In [ ]:
wanted=['13','63','157','126','22','9']
selectedAllkeysSE=allkeysSE[allkeysSE.merchant.isin(wanted) == True]
selectedAllkeysSE

id_x                                       affiliateUrl  isActive  \
3  134652039  https://www.instant-gaming.com/en/12619-/?igr=...      True   
4  134569366  https://www.cdkeys.com/hitman-3-pc-steam?mw_ar...      True   

  merchant edition region    stock platform price.eur.bestCoupon.code  \
3       13       1      2  InStock    steam                       NaN   
4        9       1      2  InStock    steam                       NaN   

  price.eur.bestCoupon.discountValue price.eur.bestCoupon.discountStrategy  \
3                                NaN                                   NaN   
4                                NaN                                   NaN   

  price.eur.bestCoupon.isCashback currency  price  price.eur.bestCoupon  \
3                             NaN      eur  25.75                   NaN   
4                             NaN      eur  25.79                   NaN   

             name  priceSEK  
3  Instant Gaming    281.97  
4      CDKeys.com    282.41

In [ ]:
allkeysSE[allkeysSE.priceSEK==allkeysSE.priceSEK.min()]

id_x                                       affiliateUrl  isActive  \
0  134513555  https://www.eneba.com/steam-hitman-3-pc-steam-...      True   

  merchant edition region    stock platform price.eur.bestCoupon.code  \
0      272       1      2  InStock    steam                  8%: AKS8   

  price.eur.bestCoupon.discountValue price.eur.bestCoupon.discountStrategy  \
0                               8.00                                     %   

  price.eur.bestCoupon.isCashback currency  price  price.eur.bestCoupon  \
0                           False      eur  19.98                   NaN   

    name  priceSEK  
0  Eneba    218.78

In [ ]:
allkeysSE.loc[allkeysSE['priceSEK'].idxmin()]

id_x                                                                             134513555
affiliateUrl                             https://www.eneba.com/steam-hitman-3-pc-steam-...
isActive                                                                              True
merchant                                                                               272
edition                                                                                  1
region                                                                                   2
stock                                                                              InStock
platform                                                                             steam
price.eur.bestCoupon.code                                                         8%: AKS8
price.eur.bestCoupon.discountValue                                                    8.00
price.eur.bestCoupon.discountStrategy                                                    %

In [ ]:
allkeysSE.nsmallest(2,'priceSEK').filter(['name','priceSEK'],axis=1)

name  priceSEK
0           Eneba    218.78
3  Instant Gaming    281.97

In [ ]:
allkeysSE.loc[allkeysSE['platform']=='steam',['priceSEK','name']].sort_values(by='priceSEK').head(3)

priceSEK            name
0    218.78           Eneba
3    281.97  Instant Gaming
4    282.41      CDKeys.com

In [ ]:
allkeysSE[allkeysSE.price==allkeysSE.price.min()]

id_x                                       affiliateUrl  isActive  \
0  134513555  https://www.eneba.com/steam-hitman-3-pc-steam-...      True   

  merchant edition region    stock platform price.eur.bestCoupon.code  \
0      272       1      2  InStock    steam                  8%: AKS8   

  price.eur.bestCoupon.discountValue price.eur.bestCoupon.discountStrategy  \
0                               8.00                                     %   

  price.eur.bestCoupon.isCashback currency  price  price.eur.bestCoupon  \
0                           False      eur  19.98                   NaN   

    name  priceSEK  
0  Eneba    218.78

-------- Start Steam ------

In [ ]:
steamAppsResponse = requests.get(
    "https://api.steampowered.com/ISteamApps/GetAppList/v2/")

steamApps = steamAppsResponse.json()
steamApps = steamApps['applist']['apps']  # ['app']
#steamApps

In [ ]:
steam_df=pd.json_normalize(steamApps)
#steam_df.set_index('appid',inplace=True)

steam_df.tail()

appid              name
154341  2064650  Tower of Fantasy
154342  1579020            Golfie
154343  1542550      SCRAP RIDERS
154344   218620          PAYDAY 2
154345   386180          Crossout

In [ ]:
import pickle

picklepath = "pickledf.pkl";

steam_df.to_pickle(picklepath)

In [ ]:
other_df=pd.read_pickle(picklepath)
other_df

appid              name
0       1897482                  
1       2112761                  
2       2016512                  
3       1820332                  
4       1360782                  
...         ...               ...
154341  2064650  Tower of Fantasy
154342  1579020            Golfie
154343  1542550      SCRAP RIDERS
154344   218620          PAYDAY 2
154345   386180          Crossout

[154346 rows x 2 columns]

------remove all empty names rows from game list-----

In [ ]:
steam_df.replace('', np.nan, inplace=True)
steam_df.dropna(subset=['name'], inplace=True)
steam_df.head()

appid                            name
6    216938   Pieterw test app76 ( 216938 )
7    660010                           test2
8    660130                           test3
24  1806870  Amber Time Resource package 01
25  1806890  Amber Time Resource package 02

-------- filter for specific game ---------

In [ ]:
#gameName='Syberia: The World Before'

In [ ]:
#exact match
steamGamedf=steam_df.loc[steam_df['name'].str.lower().eq(gameName.lower())]
if steamGamedf.empty:
    steamGamedf=steam_df.loc[steam_df['name'].str.lower().eq(gameNameAlt.lower())]
if steamGamedf.empty:
    #contains match
    steamGamedf=steam_df.loc[steam_df['name'].str.contains(gameName, case=False)]
if steamGamedf.empty:
    #contains match
    steamGamedf=steam_df.loc[steam_df['name'].str.contains(gameNameAlt, case=False)]
steamGamedf

appid      name
47360  1659040  HITMAN 3

---- make copy with url inserted------

In [ ]:
steamGamedf['url'] = steamGamedf.agg('https://store.steampowered.com/api/appdetails?appids={0[appid]}'.format, axis=1)
steamGamedf

C:\Users\robbo\AppData\Local\Temp\ipykernel_8252\3546206248.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  steamGamedf['url'] = steamGamedf.agg('https://store.steampowered.com/api/appdetails?appids={0[appid]}'.format, axis=1)


appid      name                                                url
47360  1659040  HITMAN 3  https://store.steampowered.com/api/appdetails?...

In [ ]:
#steamGamedf['url'].values[0]

In [ ]:
frames=[]
for index, row in steamGamedf.iterrows():
    #print(row["name"], row["url"])
    steamResponse = requests.get(row['url']).json() 
    if steamResponse[str(row["appid"])]['success']:
        datac=steamResponse[str(row["appid"])]['data']
        #if "price_overview" in datac:
        steamSdf=pd.json_normalize(datac)
        steamSdf.keys()
        frames.append(steamSdf)
        
single = pd.concat(frames, ignore_index=True)
single

type      name  steam_appid required_age  is_free controller_support  \
0  game  HITMAN 3      1659040           18    False               full   

                                                 dlc  \
0  [1829591, 1829592, 1829590, 1843460, 1829595, ...   

                                detailed_description  \
0  <h1>Standard Edition</h1><p><img src="https://...   

                                      about_the_game  \
0  <u><strong>WORLD OF ASSASSINATION</strong></u>...   

                                   short_description  ...  \
0  Death Awaits. Agent 47 returns in HITMAN 3, th...  ...   

                                      metacritic.url recommendations.total  \
0  https://www.metacritic.com/game/pc/hitman-3?ft...                  5260   

  achievements.total                           achievements.highlighted  \
0                 83  [{'name': 'The Result of Previous Training', '...   

  release_date.coming_soon release_date.date  \
0                    False      20 Jan, 2022   

                     support_info.url support_info.email  \
0  https://www.ioi.dk/player-support/                      

  content_descriptors.ids                          content_descriptors.notes  
0                  [2, 5]  This Game may contain content not appropriate ...  

[1 rows x 51 columns]

In [ ]:
single.keys()

Index(['type', 'name', 'steam_appid', 'required_age', 'is_free',
       'controller_support', 'dlc', 'detailed_description', 'about_the_game',
       'short_description', 'supported_languages', 'header_image', 'website',
       'legal_notice', 'developers', 'publishers', 'demos', 'packages',
       'package_groups', 'categories', 'genres', 'screenshots', 'movies',
       'background', 'background_raw', 'pc_requirements.minimum',
       'pc_requirements.recommended', 'mac_requirements.minimum',
       'mac_requirements.recommended', 'linux_requirements.minimum',
       'linux_requirements.recommended', 'price_overview.currency',
       'price_overview.initial', 'price_overview.final',
       'price_overview.discount_percent', 'price_overview.initial_formatted',
       'price_overview.final_formatted', 'platforms.windows', 'platforms.mac',
       'platforms.linux', 'metacritic.score', 'metacritic.url',
       'recommendations.total', 'achievements.total',
       'achievements.highlighted

In [ ]:
if 'price_overview.final' in single.columns:
    steamSE = single.assign(priceSEK=single["price_overview.final"]/100*rate_sek)
    #steamSE.round({'priceSEK': 1})
    steamSE.priceSEK=steamSE.priceSEK.round()
else:
    print('No price')
steamSE

type      name  steam_appid required_age  is_free controller_support  \
0  game  HITMAN 3      1659040           18    False               full   

                                                 dlc  \
0  [1829591, 1829592, 1829590, 1843460, 1829595, ...   

                                detailed_description  \
0  <h1>Standard Edition</h1><p><img src="https://...   

                                      about_the_game  \
0  <u><strong>WORLD OF ASSASSINATION</strong></u>...   

                                   short_description  ...  \
0  Death Awaits. Agent 47 returns in HITMAN 3, th...  ...   

  recommendations.total achievements.total  \
0                  5260                 83   

                            achievements.highlighted release_date.coming_soon  \
0  [{'name': 'The Result of Previous Training', '...                    False   

  release_date.date                    support_info.url support_info.email  \
0      20 Jan, 2022  https://www.ioi.dk/player-support/                      

  content_descriptors.ids                          content_descriptors.notes  \
0                  [2, 5]  This Game may contain content not appropriate ...   

  priceSEK  
0    263.0  

[1 rows x 52 columns]

In [ ]:
steamSE.filter(['name','type','controller_support',
       'price_overview.final',
       'price_overview.discount_percent',
       'price_overview.final_formatted','priceSEK'],axis=1)

name  type controller_support  price_overview.final  \
0  HITMAN 3  game               full                  2399   

   price_overview.discount_percent price_overview.final_formatted  priceSEK  
0                               60                         23,99€     263.0

In [ ]:
print(allkeysSE.loc[allkeysSE['platform']=='steam',['priceSEK','name','affiliateUrl']].values[0])

[218.78 'Eneba'
 'https://www.eneba.com/steam-hitman-3-pc-steam-key-global?af_id=allkeyshop']


In [ ]:
print(allkeysSE.loc[allkeysSE['platform']=='steam','priceSEK'].values[0])

218.78


In [ ]:
steamSE['priceSEK'].values[0]

263.0

In [ ]:
print(f"Price for {gameName} is Steam: {steamSE['priceSEK'].values[0]} AllKeyShop: {allkeysSE.loc[allkeysSE['platform']=='steam','priceSEK'].values[0]}")

Price for Hitman 3 is Steam: 263.0 AllKeyShop: 218.78


In [ ]:
print(f"{min(steamSE['priceSEK'].values[0],allkeysSE.loc[allkeysSE['platform']=='steam','priceSEK'].values[0])}")

218.78


In [ ]:
allkeycheapest=selectedAllkeysSE.loc[selectedAllkeysSE['platform']=='steam',['priceSEK','name','affiliateUrl']].sort_values(by='priceSEK').head(3)
steamPrice=steamSE['priceSEK'].values[0]
if not allkeycheapest.empty:
    allkeysPrice=allkeycheapest['priceSEK'].values[0]
    allkeysVendor=allkeycheapest['name'].values[0]
    allkeysVendorUrl=allkeycheapest['affiliateUrl'].values[0].rsplit('?af_id=allkeyshop', 1)[0]

    if allkeysPrice<steamPrice:
        print(f"Allkeys have best price at {allkeysPrice} SEK from {allkeysVendor} ({allkeysVendorUrl}) compared to Steam: {steamPrice} SEK")
    else:
        print(f"Steam is cheapest at {steamPrice} SEK")
else:
    print(f"Steam is the only option at {steamPrice} SEK")

Steam is cheapest at 263.0 SEK
